ghcide/src/Development/IDE/Core/PluginUtils.hs

```diff
@@ -0,0 +1,108 @@
+import           Data.Functor.Identity
+import qualified Data.Text                            as T
+import           Development.IDE.Core.PositionMapping
+import           Development.IDE.Core.Shake           (IdeAction, IdeRule,
+                                                       IdeState (shakeExtras),
+                                                       mkDelayedAction,
+                                                       shakeEnqueue)
+import qualified Development.IDE.Core.Shake           as Shake
+import           Development.IDE.GHC.Orphans          ()
+import           Development.IDE.Graph                hiding (ShakeValue)
+import           Development.IDE.Types.Location       (NormalizedFilePath)
+import qualified Development.IDE.Types.Location       as Location
+import qualified Development.IDE.Types.Logger         as Logger
+import qualified Ide.PluginUtils                      as PluginUtils
+import qualified Language.LSP.Types                   as LSP
+import           Prettyprinter
+import           Prettyprinter.Render.Text            (renderStrict)
+
+-- ----------------------------------------------------------------------------
+-- Plugin Error wrapping
+-- ----------------------------------------------------------------------------
+
+data GhcidePluginError
+    = forall a . Show a => FastRuleNotReady a
+    | forall a . Show a => RuleFailed a
+    | CoreError PluginUtils.PluginError
+
+instance Pretty GhcidePluginError where
+  pretty = \case
+    FastRuleNotReady rule -> "FastRuleNotReady:" <+> viaShow rule
+    RuleFailed rule       -> "RuleFailed:" <+> viaShow rule
+    CoreError perror      -> pretty $ PluginUtils.prettyPluginError perror
+
+pluginResponse ::
+    Monad m =>
+    ExceptT GhcidePluginError m a ->
+    m (Either LSP.ResponseError a)
+pluginResponse = PluginUtils.pluginResponse' handlePluginError
+
+withPluginError :: Functor m => ExceptT PluginUtils.PluginError m a -> ExceptT GhcidePluginError m a
+withPluginError = PluginUtils.withError CoreError
+
+mkPluginErrorMessage :: T.Text -> GhcidePluginError
+mkPluginErrorMessage = CoreError . PluginUtils.mkPluginErrorMessage
+
+handlePluginError :: GhcidePluginError -> LSP.ResponseError
+handlePluginError msg = PluginUtils.mkSimpleResponseError $ renderStrict simpleDoc
+    where
+        simpleDoc = layoutPretty defaultLayoutOptions $ pretty msg
+
+-- ----------------------------------------------------------------------------
+-- Action wrappers
+-- ----------------------------------------------------------------------------
+
+runAction :: MonadIO m => String -> IdeState -> ExceptT e Action a -> ExceptT e m a
+runAction herald ide act =
+  PluginUtils.hoistExceptT . ExceptT $
+    join $ shakeEnqueue (shakeExtras ide) (mkDelayedAction herald Logger.Debug $ runExceptT act)
+
+-- | Request a Rule result, it not available return the last computed result which may be stale.
+--   Errors out if none available.
```

Comment:
Can I request that we not use "errors out", it's very ambiguous when we have both `ExceptT` and exceptions in the picture. Maybe just "throws an exception"?